docs/src/main/mdoc/datatypes/statet.md

```diff
@@ -0,0 +1,244 @@
+
+The whole game state machine can be summarized as follows:
+
+```asciidoc
+
+    +----------------------+                 +----------------------+                   +-------------------+
+    | Ask for Target Word  +---------------->| Show the word hidden |------------------>| Ask for the guess |
+    +----------------------+                 +----------------------+                   +-------------------+
+                                                    ^                                             |
+                                                    |                                             |
+                                                    |                                             |
+                  +-----------+          +---------------+       +-----------------------+        |
+                  | Game Over |<---------+ Check the win |<------| Compute the new state |<-------+
+                  +-----------+          |   condition   |       +-----------------------+
+                                         +---------------+
+
+```
+
+At each step, the state might be updated and, since some I/O
+operations need to be performed along the way, a specific effect (`IO`
+monad) has to be used.
+
+We can model the game state as follow:
+
+```scala mdoc:silent
+import cats.data.StateT
+import cats.Eval
+import scala.io.StdIn._
+
+final case class GameState(
+    target: String,
+    guessedChars: Set[Char],
+    attemptNum: Int
+)
+
+sealed trait GameOver
+case object Win extends GameOver
+case object Loose extends GameOver
+
+object GameState {
+
+  val maxAttempts: Int = 3
+
+  def apply(target: String): GameState = GameState(
+    target = target,
+    guessedChars = Set.empty[Char],
+    attemptNum = 0
+  )
+
+  def showWordHidden(state: GameState): String =
+    state.target.map(c => if (state.guessedChars.contains(c)) c else '*')
+
+  def attemptGuess(
+      state: GameState,
+      guess: String
+  ): Either[GameOver, GameState] =
+    state match {
+      case GameState(t, _, _) if t == guess => Left(Win)
+      case GameState(_, _, attemptNum) if attemptNum == (maxAttempts - 1) =>
+        Left(Loose)
+      case GameState(target, guessedChars, attemptNum) =>
+        Right(
+          GameState(
+            target = target,
+            guessedChars = guessedChars ++ guess,
+            attemptNum = attemptNum + 1
+          )
+        )
+    }
+}
+```
+
+In the code above you can see some useful functions, such as the one
+that returns the encrypted target word.
+
+In the following snippet you can see the above code in action:
+
+```scala mdoc
+val gameState1 = GameState("cats")
+println(GameState.showWordHidden(gameState1))
+val gameState2 = GameState.attemptGuess(gameState1, "bats")
+println(GameState.showWordHidden(gameState2.getOrElse(???)))
+```
+
+Unfortunately, the rest of the example can't be shown here due to the
+required `cats-effect` dependency. We recommend to check out the rest
```

Comment:
Would you consider contributing this example to the Cats Effect docs? :) I think it would be cool!

Reply:
I agree. I looked quickly at the `cats-effects` doc site, but it's not structured in the same way as this one. I struggle to identify where we can put general `IO` examples like the hangman one. 

What I would like more is to extend the `example` folder in the repo(or have a separate repo) with more use cases like this one, so newcomers can go, grab the code or just check it out easily. This also applies to `cats` repo actually, but this is more of a _Megathread_ 😅

Reply:
That's true, I guess we'd have to add an "examples" section to the cats-effect docs. Hum.

Yes, this is totally a mega-thread issue :) Probably we need a unified Typelevel docs site + examples repo, that can show examples using multiple projects together, like cats and cats-effect. There are already a few projects for this purpose, like https://github.com/djspiewak/kitteh-redis.

I know the typelevel.org website is currently being redesigned and I had some ideas in https://github.com/typelevel/governance/discussions/28 that maybe we can raise some funding for improving documentation as well. Feel free to comment there or raise an issue about this topic since you seem passionate about it :)

Reply:
I'll definitely have a look there 😉 

Passionate...I'm just a lazy coder who likes to copy-paste code 😝 That's why I like the `examples` repo idea